R/ms-sql.R

```diff
@@ -0,0 +1,227 @@
+    #' @param algo for `{memoise}` compatibility, the `digest()` algorithm
+    #' @param compress for `{memoise}` compatibility, should the data be compressed?
+    #' @return A cache_postgres object
+    initialize = function(...,
+                          cache_table = "bankrcache",
+                          algo = "sha512",
+                          compress = FALSE) {
+      if (!requireNamespace("odbc")) {
+        stop(
+          paste(
+            "The {odbc} package has to be installed before using `cache_ms_sql`.",
+            "Please install it first, for example with install.packages('odbc').",
+            sep = "\n"
+          )
+        )
+      }
+      if (!requireNamespace("DBI")) {
+        stop(
+          paste(
+            "The {DBI} package has to be installed before using `cache_ms_sql`.",
+            "Please install it first, for example with install.packages('DBI').",
+            sep = "\n"
+          )
+        )
+      }
+      private$interface <- DBI::dbConnect(
+        odbc::odbc(),
+        ...
+      )
+
+      private$cache_table <- cache_table
+
+      if (
+        cache_table %in% DBI::dbListTables(private$interface)
+      ) {
+        res <- DBI::dbGetQuery(
+          private$interface,
+          sprintf(
+            "SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = '%s';",
+            cache_table
+          )
+        )
+        names(res) <- tolower(names(res))
+        attempt::stop_if_not(
+          nrow(res) == 2,
+          msg = "Your cache_table your only have two columns"
+        )
+        attempt::stop_if_not(
+          all(c("cache", "id") %in% res$column_name),
+          msg = "Your cache_db should have a `cache` and an `id` column."
+        )
+        attempt::stop_if_not(
+          all(c("varchar", "varbinary") %in% res$data_type),
+          msg = "Your cache_table data types should be `varchar` and `varbinary`."
+        )
+      } else {
+        DBI::dbCreateTable(
+          private$interface,
+          cache_table,
+          fields = c(
+            id = "varchar(max)",
+            cache = "varbinary(max)"
+          )
+        )
+      }
+
+      private$algo <- algo
+      private$compress <- compress
+    },
+    #' @description
+    #' Does the cache contains a given key?
+    #' @param key Name of the key.
+    #' @return TRUE/FALSE
+    has_key = function(key) {
+      res <- DBI::dbGetQuery(
+        private$interface,
+        sprintf(
+          "SELECT id FROM %s WHERE id = '%s';",
+          private$cache_table,
+          key
+        )
+      )
+      if (nrow(res) > 1) {
+        stop("Corrupted cache: more than one entry for ", key)
+      }
+      nrow(res) == 1
+    },
+    #' @description
+    #' Get a key from the cache
+    #' @param key Name of the key.
+    #' @return The value stored using the `key`
+    # Inspied by @jrosell https://stackoverflow.com/a/70288183/8236642
+    get = function(key) {
```

Comment:
`ms-sql$get` is the same as `postgres$get` so there is no need to redefine here (will be passed through inheritance).